src/passes/JsCallIndirect.cpp

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2017 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//
+// This pass replaces all call_indirect instructions with a call to JS side
+// js_call_indirect_xxxx functions that can perform debug analysis of the
+// call in question. Must run after FuncCastEmulation pass, if that is
+// desired.
+//
+
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+
+#include <ir/element-utils.h>
+#include <ir/literal-utils.h>
+#include <pass.h>
+#include <wasm-builder.h>
+#include <wasm.h>
+#include <asm_v_wasm.h>
+
+namespace wasm {
+
+void setFunctionAsImported(std::unique_ptr<Function>& import, Module* m);
+
+typedef std::unordered_map<std::string, std::unique_ptr<Function>>
+  JsCallIndirectMap;
+
+static std::string getCallIndirectFromArgs(CallIndirect* ci) {
+  std::string jsCallIndirectStr = std::string("js_call_indirect_");
+  jsCallIndirectStr += getSig(ci->type);
+  for (Expression*& operand : ci->operands) {
+    jsCallIndirectStr += getSig(operand->type);
+  }
```

Comment:
I think you can do `return "js_call_indirect_" + getSigFromStructs(ci->type, ci->operands)` here but I'm not sure.

Reply:
Nice, that worked.